# Lab 2 : Atanasio Giuseppe s300733

# What I did
- I've used a tuple of tuples as a genome, with each gene generated by the `problem()` function taken by lab1;
- I've tweaked the `OFFSPRING_SIZE` in order to increase it arbitrarily as `POPULATION_SIZE`. The offspring follows this rule: if problem size is greater than 50, it acts as a percentage based on 0.02 and the rounded logarithm of the problem size;

## Main methods
- `w()`: retrieves the genome's weight based on the cardinality of the alleles;
- `tournament()`: takes a random individual based on the tournament size;
- `cross_over()`: choose a cutting point based on the minimum of the two genomes' length, then creates a new mixed genome based on the cutting point and if its length is less than N then inserts a random gene from `genes` which is not inside the new gene;
- `distance()`: utility of the `fitness()`Â one;
- `fitness()`: scales over `-weight(genome)` since we want to maximize it and giving a penalty of `-MAX_INT` to the genoms which does not respect the `is_valid()` condition;
- `is_valid()`: retrieves a boolean based on `distance()` function;
- `mutation()`: choose a random cutting point in which a new random genes not present in the previous genome will be inserted inside it.

# Results

|**Problem size**|**Population Size**|**Offspring size**|**weight**|
|---|---|---|---|
|5|100|5|5|
|10|200|10|12|
|20|200|20|27|
|50|500|50|95|
|100|1000|86|241|
|500|2500|410|1690|
|1000|5000|400|4016|

## What can be done?

Surely the solution for Problem size 1000 could be improved by tweaking population and offspring size.